### PR TITLE
[cart-4]: GetShoppingSession endpoint was implemented

### PR DESCRIPTION
**Pull request by Sunagatov**
**1. Task's name:** 
[cart-4] Implement Cart API: Get the shopping session

**2. Developer's full name:** 
Zufar Sunagatov

**3. Link to the PR:** 
https://github.com/Sunagatov/Online-Store/compare/feature/cart-4/get-shopping-session?expand=1

**4. Link to Documentation:** 
https://github.com/Sunagatov/Online-Store/wiki/API-Specification-for-Cart#http-get-apiv1cartshoppingsessionid

**5. Changes:**
- `GetShoppingSession` endpoint was added.
- `findShoppingSessionByUserId` method was added to `ShoppingSessionRepository`.
- `n+1` issue was fixed for `ShoppingSession` and `ShoppingSessionItem` entities.
- Spring boot maven version was updated.
- `ShoppingSessionNotFoundException(final UUID userId)` constructor was added.
- `ShoppingSessionItem.version` type issue was fixed.
- `ShoppingSession` was removed from `ShoppingSessionItemDto`.
- `ShoppingSessionItemDtoConverter.toEntity` was deleted.
- `ShoppingSessionDtoConverter.toEntity` was deleted.
- `ShoppingSessionProvider` class was added.

**6. Tests:**
Tested with Postman by making HTTP GET request to GetShoppingSession endpoint to get the shopping session for the user with the userId = '11111111-1111-1111-1111-111111111111'

**7. Test data:**
HTTP GET request
http://localhost:8083/api/v1/cart?userId=11111111-1111-1111-1111-111111111111

**8. Conclusion:**
Project was built with Apache Maven successfully.
Project was tested manually with Postman successfully.

### CODE REVIEW

**Review by UladzislauM — COMMENTED** [review]
Reviewed.

**Review by Sanya124 — COMMENTED** [review]
Reviewed

File: src/main/java/com/zufar/onlinestore/cart/endpoint/CartEndpoint.java [hunk 1]
```diff
@@ -25,6 +29,16 @@ public class CartEndpoint {
 
     private final CartApi cartApi;
 
+    @GetMapping
+    @ResponseBody
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/cart/endpoint/CartEndpoint.java:33]: - response body

File: src/main/java/com/zufar/onlinestore/cart/endpoint/CartEndpoint.java [hunk 2]
```diff
@@ -25,6 +29,16 @@ public class CartEndpoint {
...
+    public ResponseEntity<ShoppingSessionDto> getShoppingSession(@RequestParam final String userId) {
+        log.info("Received the request to get the shoppingSession for the user with id: {}", userId);
+        ShoppingSessionDto shoppingSessionDto = cartApi.getShoppingSession(UUID.fromString(userId));
+        log.info("The shoppingSession for the user with id: {} was retrieved successfully", shoppingSessionDto.userId());
+        return ResponseEntity.ok()
+                .body(shoppingSessionDto);
+    }
+
     @PatchMapping
     @ResponseBody
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/cart/endpoint/CartEndpoint.java:43]: - response body

File: src/main/java/com/zufar/onlinestore/cart/entity/ShoppingSession.java [hunk 3]
```diff
@@ -33,7 +34,7 @@ public class ShoppingSession {
     @Column(name = "user_id", nullable = false)
     private UUID userId;
 
-    @OneToMany(mappedBy = "shoppingSession", cascade = CascadeType.ALL)
+    @OneToMany(mappedBy = "shoppingSession", cascade = CascadeType.ALL, fetch = FetchType.EAGER)
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/cart/entity/ShoppingSession.java:37]: avoid cascade type all

Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/cart/entity/ShoppingSession.java:37]: eager pls check if needed

File: src/main/java/com/zufar/onlinestore/cart/entity/ShoppingSessionItem.java [hunk 4]
```diff
@@ -25,7 +32,7 @@ public class ShoppingSessionItem {
 
     @JsonIgnore
```
Comment (yevr19) [inline src/main/java/com/zufar/onlinestore/cart/entity/ShoppingSessionItem.java:33]: remove

**Review by yevr19 — APPROVED** [review]

**Review by Sanya124 — APPROVED** [review]

File: src/main/java/com/zufar/onlinestore/cart/api/CartApi.java [hunk 5]
```diff
@@ -1,23 +1,24 @@
...
 
     /**
      * Enables to get ShoppingSession (the cart details)
      *
-     * @param getShoppingSessionRequest the request to get the shopping session (the cart details)
+     * @param userId is the identifier of the user for whom the shopping session is returned
      * @return ShoppingSessionDto (the cart details)
      * */
-    ShoppingSessionDto getShoppingSession(final GetShoppingSessionRequest getShoppingSessionRequest);
+    ShoppingSessionDto getShoppingSession(final UUID userId) throws ShoppingSessionNotFoundException;
```
[the remark this answers is not in the document]
Comment (Sunagatov) [reply src/main/java/com/zufar/onlinestore/cart/api/CartApi.java:21]: Fixed

File: src/main/java/com/zufar/onlinestore/cart/endpoint/CartEndpoint.java (hunk 1 above)
Comment (Sunagatov) [reply src/main/java/com/zufar/onlinestore/cart/endpoint/CartEndpoint.java:33]: Fixed

File: src/main/java/com/zufar/onlinestore/cart/endpoint/CartEndpoint.java (hunk 2 above)
Comment (Sunagatov) [reply src/main/java/com/zufar/onlinestore/cart/endpoint/CartEndpoint.java:43]: Fixed

File: src/main/java/com/zufar/onlinestore/cart/entity/ShoppingSession.java (hunk 3 above)
Comment (Sunagatov) [reply src/main/java/com/zufar/onlinestore/cart/entity/ShoppingSession.java:37]: Fixed
Added //TODO to think about it better

Comment (Sunagatov) [reply src/main/java/com/zufar/onlinestore/cart/entity/ShoppingSession.java:37]: Add //TODO to think about it better

File: src/main/java/com/zufar/onlinestore/cart/entity/ShoppingSessionItem.java (hunk 4 above)
Comment (Sunagatov) [reply src/main/java/com/zufar/onlinestore/cart/entity/ShoppingSessionItem.java:33]: Fixed

**Pull request merged**